File: number-of-dice-rolls-with-target-sum.cc
const int MOD = 1e9 + 7;

class Solution {
public:
  int numRollsToTarget(int n, int k, int target) {
    vector<vector<int>> dp(2, vector<int>(target + 1));
    dp[0][0] = 1;

    for (int l = 1; l <= n; ++l) {
      int x = l % 2;
      for (int v = 0; v <= target; ++v) dp[x][v] = 0;
      for (int i = 1; i <= k; ++i) for (int v = i; v <= target; ++v) {
        dp[x][v] = (dp[x][v] + dp[1 - x][v - i]) % MOD;
      }
    }

    return dp[n % 2][target];
  }
};

File: n/number-of-dice-rolls-with-target-sum.rs
const MOD: i32 = 1_000_000_007;

impl Solution {
  pub fn num_rolls_to_target(n: i32, k: i32, target: i32) -> i32 {
    let mut d = vec![vec![0; target as usize + 1]; n as usize + 1];
    d[0][0] = 1;

    for i in 1..=n as usize {
      for t in 0..=target as usize {
        for j in 1..=k as usize {
          if t >= j {
            d[i][t] = (d[i][t] + d[i - 1][t - j]) % MOD;
          }
        }
      }
    }

    d[n as usize][target as usize]
  }
}
